apps/screenshot-diff/src/computeScreenshotDiff.ts

```diff
@@ -51,6 +66,135 @@ async function completeFile({
   }
 }
 
+/**
+ * Delete the temporary file and patch the screenshot diff
+ */
+const unlinkFileAndPatchScreenshotDiff = async (
+  props: {
+    s3ImageFile: S3ImageFile;
+    screenshotDiffId: string;
+    fields: Partial<ScreenshotDiff>;
+  },
+  trx?: TransactionOrKnex | undefined,
+) => {
+  await props.s3ImageFile.unlink();
+  await ScreenshotDiff.query(trx)
+    .findById(props.screenshotDiffId)
+    .patch(props.fields);
+};
+
+/**
+ * Manages the lock to upload the diff file
+ */
+async function uploadDiffWithLock({
+  key,
+  diffResult,
+  diffImage,
+  screenshotDiffId,
+}: {
+  key: string;
+  diffResult: ImageDiffResult;
+  diffImage: S3ImageFile;
+  screenshotDiffId: string;
+}) {
+  const lock = await getRedisLock();
+  const lockKey = `diffUpload-${key}`;
+  await lock.acquire(lockKey, async () => {
+    // Check if the diff file has been uploaded by another process
+    const existingDiffFile = await File.query().findOne({ key });
+    const patchProps = {
+      screenshotDiffId,
+      s3ImageFile: diffImage,
+      fields: { s3Id: key, score: diffResult.score },
+    };
+
+    if (existingDiffFile) {
+      await unlinkFileAndPatchScreenshotDiff({
+        ...patchProps,
+        fields: { ...patchProps.fields, fileId: existingDiffFile.id },
+      });
+      return;
+    }
+
+    await diffImage.upload();
+    await transaction(async (trx) => {
+      const diffFile = await File.query(trx)
+        .insert({ key, width: diffResult.width, height: diffResult.height })
+        .returning("*");
+      await unlinkFileAndPatchScreenshotDiff(
```

Comment:
No need to do that in the lock, only the file creation. Updating the diff should not be done in the lock. We don't care about the transaction here.